cloud/cluster/cluster.go

```diff
@@ -1,51 +1,121 @@
 package cluster
 
 import (
-	"fmt"
-	"io"
+	"sort"
 )
 
-// NodeUpdate represents a change to the cluster
-type NodeUpdate struct {
-	UpdateType NodeUpdateType
-	Id         NodeId
-	Node       Node // Only set for adds
+// Cluster represents a cluster of Nodes.
+
+type Cluster struct {
+	State    *State
+	reqCh    chan interface{}
+	updateCh chan []NodeUpdate
+	StateCh  chan []Node
+	subs     []chan []NodeUpdate
 }
 
-func (u *NodeUpdate) String() string {
-	return fmt.Sprintf("%v %v %v", u.UpdateType, u.Id, u.Node)
+func NewCluster(state []Node, updateCh chan []NodeUpdate, stateCh chan []Node) *Cluster {
```

Comment:
I mean, you can keep both, but I shouldn't have to read the impl to know that I can use nil for one of the channels. You could also make NewCluster private and replace it with one constructor for each update type (that's not necessarily better though).